OCaml/sudoku_solver/lib/sudoku.ml

```diff
@@ -0,0 +1,75 @@
+(* OCaml Version of the code *)
+
+module Sudoku = struct
+  (* Main function to run *)
+  type board = int array array
+
+  let print_board board = 
+    Array.iter (fun row ->
+      Array.iter (fun col -> Printf.printf "%d " col) row;
+      print_newline()) board
+  ;;
+
+  (* to check if the number is in row, col and subgrid *) 
+  let is_valid_move board row col num = 
+    Array.exists (( = ) num) board.(row) && 
+    Array.exists (fun r -> r.(col) = num) board &&
+    Array.exists (fun r -> Array.exists (( = ) num) (Array.sub r col 3)) (Array.sub board row 3)
+  ;;
+
+  (* check if each row and col if = 0, then return coordinates *)
+  let find_empty_cell board = 
+    let rec aux row col =
+      if row = Array.length board then None (* reach the end *)
+      else if col = Array.length board && row < Array.length board then aux (row + 1) col
+      else if board.(row).(col) = 0 then Some (row, col)
+      else aux row (col + 1)
+  in 
+    aux 0 0
+  ;;
+
+  let solve_sudoku board =
+    None  
```

Comment:
What I do when I write the skeleton code is to use this instead

```suggestion
  let solve_sudoku board = failwith "Not implemented"  
```

Reply:
right. I see this too with the practice. Thank you!